crates/wysiwyg/src/dom/nodes/container_node.rs

```diff
@@ -661,30 +686,56 @@ impl<S: UnicodeString> ContainerNode<S> {
         selection_writer: Option<&mut SelectionWriter>,
         state: ToHtmlState,
     ) {
+        let as_message = false;
         assert!(matches!(self.kind, ContainerNodeKind::Paragraph));
         let name = self.name();
 
         self.fmt_tag_open(name, formatter, &self.attrs);
         if self.is_empty() {
             formatter.push(char::nbsp());
         }
-        self.fmt_children_html(formatter, selection_writer, state);
+        self.fmt_children_html(formatter, selection_writer, state, as_message);
         self.fmt_tag_close(name, formatter);
     }
 
+    fn fmt_paragraph_as_message_html(
+        &self,
+        formatter: &mut S,
+        selection_writer: Option<&mut SelectionWriter>,
+        state: ToHtmlState,
+    ) {
+        let as_message = true;
+        assert!(matches!(self.kind, ContainerNodeKind::Paragraph));
+        let name = self.name();
+
+        if self.is_empty() {
+            formatter.push('\n');
```

Comment:
Format empty paragraphs as newlines, behaviour consistent with the legacy Element composer.